### ISSUE TITLE
FR: Safari Push Notification Support (via APN)

### ISSUE THREAD

**Issue by digaus** [firebase/firebase-js-sdk#1423]
Any plans on supporting the Safari Push Notifications (implement APN handling and conversion to FCM keys)? 

https://developer.apple.com/notifications/safari-push-notifications/

**Comment by timscullin** [comment]
+1

**Comment by ganglio** [comment]
+1

**Comment by prashantalhat** [comment]
+1

**Comment by avithe1** [comment]
+1

**Comment by planecore** [comment]
+1

**Comment by macosxvn** [comment]
+1

**Comment by 1hakr** [comment]
+1

**Comment by sagivo** [comment]
+1

**Comment by srobertson421** [comment]
+1

This is really needed, still doesn't work in Safari 13

**Comment by niklat22** [comment]
+1

can we use any third party notification which interact with FCM?

**Comment by davwheat** [comment]
Just wondering if there are any updates on this issue as it seems to be completely silent except from users of Firebase.

**Comment by Onyoursix** [comment]
Would love an update on this, this is sorely needed

**Comment by chukhanhvan** [comment]
+1

**Comment by LRK321** [comment]
+1

**Comment by andreasanta** [comment]
Same here, it's a big turn down for us, it [represents 27% of the Market Share on mobile](https://netmarketshare.com/browser-market-share.aspx?options=%7B%22filter%22%3A%7B%22%24and%22%3A%5B%7B%22deviceType%22%3A%7B%22%24in%22%3A%5B%22Mobile%22%5D%7D%7D%5D%7D%2C%22dateLabel%22%3A%22Trend%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22browser%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22browsersDesktop%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222019-04%22%2C%22dateEnd%22%3A%222020-03%22%2C%22segments%22%3A%22-1000%22%7D), and [3% on desktop](https://netmarketshare.com/browser-market-share.aspx?options=%7B%22filter%22%3A%7B%22%24and%22%3A%5B%7B%22deviceType%22%3A%7B%22%24in%22%3A%5B%22Desktop%2Flaptop%22%5D%7D%7D%5D%7D%2C%22dateLabel%22%3A%22Trend%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22browser%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22browsersDesktop%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222019-04%22%2C%22dateEnd%22%3A%222020-03%22%2C%22segments%22%3A%22-1000%22%7D).

Ignoring Safari on mobile is definitely not an option in our PWA.

**Comment by rejhgadellaa** [comment]
@andreasanta You mention Safari on iOS but even Apple's own APN push notifications only work on macOS currently. There is no support for push messages in Safari on iOS. No amount of money, development will change that unless Apple deems it worth the time to implement.

That said: I'd love Safari on macOS support so consider me /subbed

**Comment by andreasanta** [comment]
@rejhgadellaa So unfortunately true! I thought APN would work on Safari iOS. Apple is still determined to restrict core functionality to Apps, so they can cash in a while longer on the App Store purchases. You're right, nothing can change that unless they open the functionality to PWAs.

**Comment by malarium** [comment]
YUP, lack of iOS support truly sucks. But I'm pretty sure that when it is finally introduced by Apple - it will be implemented in FCM, wont it?

**Comment by TraMZzz** [comment]
+

**Comment by AverageHelper** [comment]
I'm sure iOS support is coming.  Meantime, macOS support would be nice.

**Comment by mourad1081** [comment]
+1

**Comment by phips28** [comment]
+1

**Comment by lukepighetti** [comment]
Wish Google would stop playing Push API chicken with WebKit and start supporting macOS Safari with Firebase Cloud Messaging. We all know Apple isn't going to do anything about this. It's in Google's court. They have an implementation green light.

**Comment by lukepighetti** [comment]
I'm at a bit of a loss here. I did what I thought was my due diligence to recommend FCM to a client. Only after finding the `messaging/unsupported-browser` error in console did I realize that macOS Safari, despite Google having an implementation green light, is not supported by FCM.

Checking the WebKit feature roadmap there are no expectations of allowing Push API. This is 100% in Google's court to implement.

We are already fully prepared to tool up for APN certs for our mobile app, I see no reason why we, as Firebase customers, should be expected to ignore macOS Safari.

**Comment by zwu52** [comment]
Hey folks,
Thanks for expressing great interest in having FM supporting MacOS Safari sends.
Apologies for the negligence on this long standing FR. I have brought this up to the team and want to share some updates: 

- FM team has been committing to an internal infrastructure improvement program. Limited capacity is currently dedicated to new features.
- This FR is in the front queue of the next to be implemented.
- We are pushing to reprioritize this feature to fit into the infra program. 
 
I'll keep the thread posted on any updates. 
Cheers.

**Comment by malarium** [comment]
At last! Thank you and, well, let's wait some more ;)

**Comment by digaus** [comment]
Good to see that we are finally making progress 🎉

**Comment by neekey** [comment]
OK, given the push service is not available yet, is there anyway i can only use the send / subscribe feature while user's page is open? I am atm only want to use FCM as a service for service to send connected client data update event. The missing of safari support becomes the major support for me

**Comment by zwu52** [comment]
@neekey  I assume you are targeting Mac OS Safari. Unfortunately, until this feature is released, you will need to find another way to push data payloads to your Mac OS Safari clients. But you probably know you could already do it to other browsers with the FM SDK with the caveat that you can't silent push.

**Comment by samaneh-kamalian** [comment]
Is there any update for this issue?

**Comment by dcrwalz** [comment]
Please apply this feature.
My client doesn't eat any food beside Apple.

**Comment by AntoineGuerra** [comment]
+1 🤓

**Comment by kashyaprahul94** [comment]
+1 🤓

**Comment by zwu52** [comment]
Hi folks, thanks for the patience. update: 

- the major FCM internal infrastructure improvement program continues (planned to complete on Q3 2023). 
- We've been observing an increasing number of outages as a result.
- Therefore, the team has put more weight on reliable refactor and rollout.  
- The current  phase is focusing on refactoring the Webpush and IID services and is projected to be completed by Q4 2022. 
- FCM client team (previously driving new feature development) is currently understaffed. 
- Sadly, At this time, we can't provide a clear timeline for this FR.   😞 😿

**Comment by lukepighetti** [comment]
the transparency is.. incredible. Thank you

**Comment by CosmicPangolin** [comment]
Deeply concerning in my view - 18 month refactor, with no timeline on a core feature that's been an open problem for 3 years?  Haaard yikes, dudes.  Something internal needs to change.

**Comment by lukepighetti** [comment]
Way better to wait years for something that takes years, instead of waiting months for something that takes years. The transparency gives you what you need to make decisions, which is all we can ask for

**Comment by CosmicPangolin** [comment]
> Way better to wait years for something that takes years, instead of waiting months for something that takes years. The transparency gives you what you need to make decisions, which is all we can ask for

You're right, that was super unfair of me. This is a valuable feature but it should be done right, and systems work is worth its weight in gold over time. I hope you guys get the staff you need <3

**Comment by FirefighterBlu3** [comment]
Howdy, we're now in Q4 of 2022, @zwu52 would you mind posting another update?

**Comment by ferinoc** [comment]
@zwu52 Hi, if you could give us an update would great. Thank you :)

**Comment by noobcoder95** [comment]
i'm waiting for update

**Comment by sman591** [comment]
The [FCM quick start app](https://github.com/firebase/quickstart-js/tree/master/messaging) runs background notifications successfully for me on Safari 16.2 with macOS Ventura. macOS 13 Ventura (released Oct 2022) is the minimum requirement according to [Apple's documentation](https://developer.apple.com/documentation/usernotifications/sending_web_push_notifications_in_safari_and_other_browsers)

**Comment by huksley** [comment]
@sman591 What about iOS? Could you check it?

**Comment by lukepighetti** [comment]
Boy, was I wrong. Apple did implement Push API!

**Comment by sman591** [comment]
@huksley Apple's documentation only mentions macOS for Web Push

**Comment by barhemo** [comment]
any news?

**Comment by andreififiita** [comment]
I tried firebase messaging v9.16.0 and it seems to work in background (when delivered by the service worker) but in foreground only receives notifications 3 times, then it stops working. Is this a "normal" behavior for some reason?

**Comment by andreififiita** [comment]
Also, i tested on iOS and i don't understand how can you try the experimental Push API in Safari 16 iOS, since the Notification.requestPermission() doesn't even exist? https://developer.mozilla.org/en-US/docs/Web/API/Notification/permission

**Comment by AverageHelper** [comment]
> the Notification.requestPermission() function doesn't even exist

You mean [this](https://developer.mozilla.org/en-US/docs/Web/API/Notification/requestPermission) `Notification.requestPermission()` function?

**Comment by andreififiita** [comment]
> > the Notification.requestPermission() function doesn't even exist
> 
> You mean [this](https://developer.mozilla.org/en-US/docs/Web/API/Notification/requestPermission) `Notification.requestPermission()` function?

Exactly, iOS doesn't have that at 16.2 from what i tested. Or do we have at least a substitute for iOS?

**Comment by rejhgadellaa** [comment]
Web Push for Safari is currently only available on macOS 13.x. 

( Even though you can install the latest Safari on macOS 12, that version will not have Web Push because their implementation requires the OS update. It relies on some new system service (yes, really...) )

For iOS and iPadOS, Web Push support will come [in 2023](https://webkit.org/blog/12945/meet-web-push/#:~:text=Keep%20an%20eye%20out%20for%20Web%20Push%20on%20iOS%20and%20iPadOS%20in%202023.):
https://webkit.org/blog/12945/meet-web-push/

**Comment by andreififiita** [comment]
> Web Push for Safari is currently only available on macOS 13.x.
> 
> ( Even though you can install the latest Safari on macOS 12, that version will not have Web Push because their implementation requires the OS update. It relies on some new system service (yes, really...) )
> 
> For iOS and iPadOS, Web Push support will come [in 2023](https://webkit.org/blog/12945/meet-web-push/#:~:text=Keep%20an%20eye%20out%20for%20Web%20Push%20on%20iOS%20and%20iPadOS%20in%202023.): https://webkit.org/blog/12945/meet-web-push/

Yes, i actually connected to the ios 16 simulator Safari console and checked the JS APIs. PushManager is there (and other Push API classes) but Notification is missing (needed for Notification.requestPermission). Also navigator.permissions.query throws exception for name: "notifications". Maybe the boys at safari still need a bit more time to finish the user prompt popup for asking permissions...

**Comment by NataliaTsoy** [comment]
Hi guys. Checking the browser `firebase.messaging.isSupported()` returns `true` in Safari 16.3 macOS 13, but `messaging.getToken()` returns nothing. Does this mean that web push for Safari are not available on firebase?
After a bit of debugging, I got that Safari requires initialization of the Service Worker before requesting permission from the user. And if I initialize the Service Worker before, then the `getToken()` function updates this Service Worker and it seems to block the permission request

**Comment by josehigroup** [comment]
+1 would be awesome to get!

**Comment by MrAlexand0r** [comment]
> Hi guys. Checking the browser `firebase.messaging.isSupported()` returns `true` in Safari 16.3 macOS 13, but `messaging.getToken()` returns nothing. Does this mean that web push for Safari are not available on firebase? After a bit of debugging, I got that Safari requires initialization of the Service Worker before requesting permission from the user. And if I initialize the Service Worker before, then the `getToken()` function updates this Service Worker and it seems to block the permission request

Have you found a solution for this? I have the same Problem.
`isSupported()` returns `true` but the `getToken(...) ` Promise is never resolved. No error or anything.
Weirdly enough, this occurs only on MacOS. On iPadOS and iOS it's working as intended

**Comment by NataliaTsoy** [comment]
> Have you found a solution for this? I have the same Problem. `isSupported()` returns `true` but the `getToken(...) ` Promise is never resolved. No error or anything. Weirdly enough, this occurs only on MacOS Safari. On iPadOS and iOS Safari it's working as intended

no, no solution has been found. I did it with APNs without the participation of firebase

**Comment by mourad1081** [comment]
We're almost hald a decade later now...

**Comment by zwu52** [comment]
Circling back to the thread, https://firebase.blog/posts/2023/08/fcm-for-safari/ provides some info on how to push to ios safari

**Comment by andreififiita** [comment]
It doesn't work in Safari 17.4 anymore (tested on MacOS, but i believe iOS also won't work). It was working fine from Safari 16.3 to 17.3. I posted details here, in case anyone is experiencing the same issues: https://stackoverflow.com/questions/78323312/firebase-cloud-messaging-web-issues-in-safari-17-4
